### PR TITLE
Add Rust bloom filter

### PR DESCRIPTION
**Pull request by jonathanlofgren**
Was inspired by @Bertrand31 scala implementation

### CODE REVIEW

File: rust/data_structures/bloom_filter.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    /// Returns an approximation of the number of items
+    /// inserted into the BloomFilter. 
+    fn approx_item_count(&self) -> usize {
+        let bits_set = self.bits.iter().filter(|&b| *b).count() as f64;
+        let length = self.bits.len() as f64;
+        
+       ((-length / self.hash_functions as f64) * (1.0 - bits_set / length).ln()).round() as usize
+    }
```
Comment (Bertrand31) [inline rust/data_structures/bloom_filter.rs]: Ah nice, I shall add this to my implementation.

File: rust/data_structures/bloom_filter.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        Self { bits, hash_functions }
+    }
+
+    /// Inserts the item into the BloomFilter
+    fn insert<T: Hash>(&mut self, item: &T) {
+        for i in 0..self.hash_functions {
+            let ix = self.get_index(item, i);
+            self.bits[ix] = true;
+        }
+    }
```
Comment (Bertrand31) [inline rust/data_structures/bloom_filter.rs]: Suggestion: you could try using immutable data structures (https://github.com/bodil/im-rs)

File: rust/data_structures/bloom_filter.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        (hasher.finish() % length) as usize
+    }
+
+    /// Returns an approximation of the number of items
+    /// inserted into the BloomFilter. 
+    fn approx_item_count(&self) -> usize {
+        let bits_set = self.bits.iter().filter(|&b| *b).count() as f64;
+        let length = self.bits.len() as f64;
+        
+       ((-length / self.hash_functions as f64) * (1.0 - bits_set / length).ln()).round() as usize
```
Comment (nicholette) [inline rust/data_structures/bloom_filter.rs]: @jonathanlofgren Would be great if there's a description on what this is doing/ how this expression is derived/ link to any article 😄

File: rust/data_structures/bloom_filter.rs (hunk 2 above)
Comment (jonathanlofgren) [reply rust/data_structures/bloom_filter.rs]: Looks interesting, but I wonder how immutable data structured play with a language like Rust. Rust already has some interesting properties which seems to make mutability less of a problem to work with. Basically in rust for some memory A, you can only have one of the following at a time:
* A single mutable reference to A
* Any number of immutable references to A

This means that it is impossible to have data races for instance, and rust will complain at compile time if you don't follow these rules.

But there are probably other instances where this kind of immutable data structures can still be useful.

File: rust/data_structures/bloom_filter.rs (hunk 3 above)
Comment (jonathanlofgren) [reply rust/data_structures/bloom_filter.rs]: Yes, certainly :)

File: rust/data_structures/bloom_filter.rs (hunk 2 above)
Comment (Bertrand31) [reply rust/data_structures/bloom_filter.rs]: Oh very interesting!
Yeah aside from distributed computing and data race concerns, immutability also promotes a more "pure" style of programming, in the FP sense of the word.
Meaning that your function can only return stuff, but cannot produce side effects. Stuff in, stuff out, and that's it. It makes code much easier to read and test.

Comment (Bertrand31) [reply rust/data_structures/bloom_filter.rs]: In VIS, you'd often see stuff like this:
```
... lots of code
this.processCustomMetrics(rows, stuff, foo, bar)
... more code
```
This function doesn't return anything? What does it do? What does it operate on? It must produce side effects, but what?
This is literally made impossible by immutability.

**Pull request merged**